### ISSUE TITLE
How to add correlation in the response?

### ISSUE THREAD

**Issue by rayyzhong** [dreamhead/moco#158]
I want to create the response to client, including a a parameter which is provided in his request. 

E.g. 
When he request 1 
/foo/bar?seq=1
the reponse should be {jsondata, seq : 1}

When he request 999
/foo/bar?seq=999
the reponse should be {jsondata, seq : 999}

**Comment by dreamhead** [comment]
What you need is `query`

https://github.com/dreamhead/moco/blob/master/moco-doc/apis.md#query-parameter

**Comment by rayyzhong** [comment]
Thanks for your resp. 
Let me elaborate more that I want to intercept the value of the query parameter (dynamic), and attach it to the response body. 
The use case is for responding to a JSONP call.

**Comment by dreamhead** [comment]
@rayyzhong Good use case.

Could you please provide me a detailed example if your case works?

**Comment by rayyzhong** [comment]
Try to read through the codes but could not figure out a good implementation, need your advice on how to achieve this. @dreamhead

**Comment by dreamhead** [comment]
Show your detailed case, I'll try to figure out a solution.

**Comment by rayyzhong** [comment]
Not sure I eraborate clear enough, use case as below.
Scenario 1
Given the back-end server host is backend.server.com
And server is provided with URI /server/fund
When I request mock data with URI http://backend.server.com/service/fund?callback=angular.callbacks._0
Then I should receive text response with data
**angular.callbacks._0**({"foo":"bar"})

Scenario 2
Given the back-end server host is backend.server.com
And server is provided with URI /server/fund
When I request mock data with URI http://backend.server.com/service/fund?callback=angular.callbacks._1
Then I should receive text response with data
**angular.callbacks._1**({"foo":"bar"})

The above rules should be added by a single matching as below.
server.request(by(uri("/service/fund"))).response(String.format("%s({\"foo\":\"bar\"})"), real_time_value_of_callback_query_parameter_from_the_request);

**Comment by treppo** [comment]
I have a similar scenario, where the server needs to return a cryptographic token, that is sent with the request otherwise the response will be refused. To mock this server I was not able to use moco so far as there is no way to access the request in the response handler.
Looking at the code I saw that the response handlers do get the request passed in [AbstractHttpResponseHandler](https://github.com/dreamhead/moco/blob/d613ba519350bc0ff748b97f48e9a99eb1a990bc/moco-core/src/main/java/com/github/dreamhead/moco/handler/AbstractHttpResponseHandler.java#L10) and its subclasses. I'm not sure how it would be possible to make the request accessible in the DSL. Maybe something like this (I'm not too familiar with Java):

```java
server.response(
  new RequestAwareResponseHandler() {
    @Override
    public ResponseHandler withRequest(HttpRequest request) {
      // do something with the request and return a response handler
    }
  }
);
```

What do you think?

**Comment by treppo** [comment]
I quickly put together a ResponseHandler like I described before https://github.com/treppo/moco/commit/8ab9320481dd018770a84475b97c35b93db8ed82

**Comment by treppo** [comment]
If I understand Java 8 Lambda correctly this could be used quite nicely:

```java
server.response((request) -> {
    // e.g. with a request to /path?foo=bar
    return with(request.getQueries().get("foo")[0]);
});
```

**Comment by dreamhead** [comment]
@rayyzhong Maybe what you need is already there. Check the following link.

https://github.com/dreamhead/moco/blob/master/moco-doc/apis.md#event

**Comment by treppo** [comment]
@dreamhead I don't see how event handling gives you access to the request.

**Comment by dreamhead** [comment]
Take the above example

```java
server.request(by(uri("/service/fund"))).response("event").on(complete(get("http://your_site/angular.callbacks._1")));
```

**Comment by treppo** [comment]
@dreamhead If I understand @rayyzhong correctly then the name of the angular callback function is not known in advance it could be `angular.callbacks._1` but it could also be `angular.callbacks._24317236986723` and you don't know that it advance. that's why you would need access to the request and use part of the request in the response.
I have a similar problem where i need to access a part of the request query parameters, that is generated (a crypto hash), and return it in the response.

**Comment by rayyzhong** [comment]
@treppo Yes you are correct. I need to intercept the **value** from the query parameter (dynamic), and attach it to the response body.
The **value** is unknown before the request reach to moco server.

**Comment by dreamhead** [comment]
@rayyzhong @treppo By default, Moco is a static server, which means you've setup everything before your server started.

In this case, one solution with current Moco is as follows:

```java
server.request(and(by(uri("/service/fund"))), query("first_query", "first")).response("event").on(complete(get("http://your_site/angular.callbacks._1")));
server.request(and(by(uri("/service/fund"))), query("second_query", "second")).response("event").on(complete(get("http://your_site/angular.callbacks._2")));
```

But I totally understand what you need, I'm going to provide a dynamic version in next version. The solution will be 

```java
server.request(by(uri("/service/fund"))).response("event").on(complete(get(template("http://your_site/${req.queries['your_query']}"))));
```

**Comment by songguohfut** [comment]
NEED HELP~
In my case, the api returns something like:`_jsonpbfyzwg21ezb({"code":"0000","msg":"success"})` where `_jsonpbfyzwg21ezb` is based on the request queries。

I know `template`, but I can't combine `template` with my own response data, Is there any solution to  solve my problem?

**Comment by dreamhead** [comment]
@songguohfut You can try to write your own code with Java like this.

```java
server.response(text((request) -> "foo"));
```
